libraries/tickfs/src/flash_controller.rs

```diff
@@ -0,0 +1,100 @@
+/// to the values in your implementation.
+///
+/// The boiler plate for an implementation will look something like this
+///
+/// ```rust
+/// use tickfs::error_codes::ErrorCode;
+/// use tickfs::flash_controller::FlashController;
+///
+/// #[derive(Default)]
+/// struct FlashCtrl {}
+///
+/// impl FlashCtrl {
+///     fn new() -> Self {
+///         Self { /* fields */ }
+///     }
+/// }
+///
+/// impl FlashController for FlashCtrl {
+///     fn read_region(&self, region_number: usize, offset: usize, buf: &mut [u8]) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+///
+///     fn write_region(&self, address: usize, buf: &[u8]) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+///
+///     fn erase_region(&self, region_number: usize) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+/// }
+/// ```
+pub trait FlashController {
+    /// This function must read the data from the flash region specified by
+    /// `region_number` into `buf`. The length of the data read should be the
+    /// same length as buf. `offset` indicates an offset into the region that
+    /// should be read.
+    ///
+    /// On success it should return nothing, on failure it
+    /// should return ErrorCode::ReadFail.
+    ///
+    /// If the read operation is to be complete asynchronously then
+    /// `read_region()` can return `ErrorCode::ReadNotReady(region_number)`.
+    /// By returning `ErrorCode::ReadNotReady(region_number)`
+    /// `read_region()` can indicate that the operation should be retried in
+    /// the future.
+    /// After running the `continue_()` functions after a async
+    /// `read_region()` has returned `ErrorCode::ReadNotReady(region_number)`
+    /// the `read_region()` function will be called again and this time should
+    /// return the data.
+    fn read_region(
+        &self,
+        region_number: usize,
+        offset: usize,
+        buf: &mut [u8],
+    ) -> Result<(), ErrorCode>;
+
+    /// This function must write the length of `buf` to the specified address
+    /// in flash.
+    /// If the length of `buf` is smaller then the minimum supported write size
+    /// the implementation can write a larger value. This should be done by first
+    /// reading the value, making the changed from `buf` and then writing it back.
+    ///
+    /// On success it should return nothing, on failure it
+    /// should return ErrorCode::WriteFail.
+    ///
+    /// If the write operation is to be complete asynchronously then
+    /// `write_region()` can return `ErrorCode::WriteNotReady(region_number)`.
+    /// By returning `ErrorCode::WriteNotReady(region_number)`
+    /// `read_region()` can indicate that the operation should be retried in
+    /// the future. Note that that region will not be written
+    /// again so the write must occur otherwise the operation fails.
+    fn write_region(&self, address: usize, buf: &[u8]) -> Result<(), ErrorCode>;
```

Comment:
This interface seems pretty hard to get right. It sounds like any failed write should be retried, but probably not retried in a loop (because the kernel is single threaded, so any bottom half interrupts needed to make TickFS ready to write will not run until the caller somehow yields). So the caller probably needs to schedule a deferred call anytime a write fails? But I don't see any use of `DeferredCall` in this PR. Retrying the call from within a deferred call will require that the buffer being written be stored across invocations of the kernel loop, which will require a buffer with a lifetime at least as long as that of the calling object (i.e. I see no way to reliably write from a stack allocated buffer with this interface, because any failure will require storing the buffer in something that outlives the stack frame of the caller).

Reply:
Yep, the interface is a little more complex then I would like it to be. This is the tradeoff to allow non `'static` buffers for other users (if there are any).

So a write doesn't have to be re-tried. A write is the last operation and nothing happens after a write. So the idea is a write will occur and then we are done. If the write actually fails in the driver, then we should get an error callback which we have to handle (I don't handle errors here yet). Eventually the write will complete in the driver, so we don't need to reschedule it.

Reply:
Oh, I think I understand, `WriteNotReady` means that the write is going to happen eventually, it just didn't happen immediately. That is more usable. I am not sure that should be reported as an error -- perhaps this could return something like `Result<SuccessCode, ErrorCode>`.
with
```rust
pub enum SuccessCode {
    Written
    Queued
}
```?

It seems weird that if TickFS is implemented on off-chip flash every single write would return an error, despite eventually succeeding.

Reply:
Yep, exactly!

I agree, it isn't really an error, but I also didn't want to report a success.

I was trying to avoid having a `SuccessCode` as I think that's a little convoluted. I'll add it in though.

Reply:
Done.